core/src/test/java/io/forestframework/core/http/HandlerIntegrationTest.java

```diff
@@ -0,0 +1,458 @@
+        response.write(" is handled by custom error handler", "UTF-8");
+        response.write("  e.getMessage():  " + e.getMessage());
+        runUnderTrace("custom500ErrorHandler");
+    }
+}
+
+@Router("/preHandler")
+class PreHandlerReturnsVoidTrueOrFalse extends AbstractHandlerIntegrationTest {
+    @PreHandler("/returns/*")
+    public boolean preHandleReturnsTrue(HttpServerRequest request, HttpServerResponse response) {
+        runUnderTrace(request.path());
+        response.write(request.path());
+        return request.path().contains("True");
+    }
+
+    @PreHandler("/returns/*")
+    public void preHandleContinuesWhenTrue(HttpServerRequest request, HttpServerResponse response) {
+        response.write(request.path());
+        runUnderTrace("prehandlerContinues");
+    }
+
+    @Route(type = RoutingType.PRE_HANDLER, order = 0, path = "/returnsVoid")
+    public void preHandlerReturnsVoid(HttpServerRequest request, HttpServerResponse response) {
+        response.write("preHandlerContinues: " + request.path());
+        runUnderTrace(request.path());
+    }
+
+    @Route(type = RoutingType.PRE_HANDLER, order = 1, path = "/returnsVoid")
+    public void preHandleContinuesWhenVoid(HttpServerRequest request, HttpServerResponse response) {
+        response.write(request.path());
+        runUnderTrace("prehandlerContinues");
+    }
+
+    @Route(path = "/returnsVoid", type = RoutingType.HANDLER)
+    public void handlerContinuesWhenPrehandlerReturnsVoid() {
+        runUnderTrace("handlerContinues");
+    }
+
+    @Route(path = "/returns/*", type = RoutingType.HANDLER)
+    public void handlerContinuesWhenPrehandlerReturnsTrue() {
+        runUnderTrace("handlerContinues");
+    }
+}
+
+@Router("/custom400error")
+class Custom400ErrorHandler extends AbstractHandlerIntegrationTest {
+
+    @Route(path = "/METHOD_NOT_ALLOWED", type = RoutingType.HANDLER)
+    public void handler405(HttpServerResponse response) {
+        runUnderTrace("METHOD_NOT_ALLOWED");
+        throw new HttpException(HttpStatusCode.METHOD_NOT_ALLOWED);
+    }
+
+    @Route(path = "/NOT_ACCEPTABLE", type = RoutingType.HANDLER)
+    public void handler406(HttpServerResponse response) {
+        runUnderTrace("NOT_ACCEPTABLE");
+        throw new HttpException(HttpStatusCode.NOT_ACCEPTABLE);
+    }
+
+    @Route(path = "/UNSUPPORTED_MEDIA_TYPE", type = RoutingType.HANDLER)
+    public void handler415(HttpServerResponse response) {
+        runUnderTrace("UNSUPPORTED_MEDIA_TYPE");
+        throw new HttpException(HttpStatusCode.UNSUPPORTED_MEDIA_TYPE);
+    }
+
+    @OnError(value = "/**")
+    public void on400(HttpServerResponse response, Throwable e) {
+        response.write(" is handled by custom 400errorhandler", "UTF-8");
+        response.write("  e.getMessage():  " + e.getMessage());
+        runUnderTrace("custom400ErrorHandler");
+    }
+}
+
+@Router("/errorInCustomErrorHandler")
+class ErrorInCustomErrorHandler2 extends AbstractHandlerIntegrationTest {
```

Comment:
Why `Handler2`? Where's `Handler1`?

Reply:
There's no `Handler1`. There's already an `ErrorInCustomErrorHandler` class in package, so i've got to rename it - adding a random number. I know it's not a good name. I'll change it.